Table calculations

harshilsharma63 opened the pull request:
#### Summary
Implemented table calculations

#### Ticket Link
Implements #461 

Calculations implemented -

1. count
1. countValue
1. countUniqueValue
1. sum
1. average
1. median
1. min
1. max
1. range

hahmadia commented:
Yo Harshil,

Before doing code review I decided to test it out. Here are some minor things I found after a quick run through. 

When the column too small, the range gets stacked on top like this ( I think it should be come ellipsis)
![image](https://user-images.githubusercontent.com/17804942/127164855-c3a4dc70-d12f-4ddf-ad75-76e15eaf03d3.png)

For date, the unique should be 3 here right?
![image](https://user-images.githubusercontent.com/17804942/127165139-fc176667-9b44-42bb-819e-215564a201a1.png)

Count values for multi select should be 7 here right?
![image](https://user-images.githubusercontent.com/17804942/127165350-f048caca-ea86-4c6c-995c-fc166ae7cab9.png)


For Checkbox, shouldn't false values be accounted for too?
![image](https://user-images.githubusercontent.com/17804942/127165536-535edc1c-ec32-4e2b-9d81-9ae97150e749.png)

Unique values here should be 2 (I assume we account for false values too if I'm not wrong)
![image](https://user-images.githubusercontent.com/17804942/127165641-fb202066-8159-4b68-ad5d-e2d10a9c93c1.png)

Created Time 0 values? Is this right?
![image](https://user-images.githubusercontent.com/17804942/127165801-49f19af0-3cec-45c6-b5b4-98c15ca1bc40.png)


Created time unique values also 0?
![image](https://user-images.githubusercontent.com/17804942/127165851-0c6b5c0b-37ce-4aa8-bd08-e403d0361459.png)

Created by value is 0?
![image](https://user-images.githubusercontent.com/17804942/127165995-6be3b0ab-780a-4a9d-9b64-1a1c233a91fb.png)


Created by unique value is also 0?
![image](https://user-images.githubusercontent.com/17804942/127166054-046bba50-42cd-4f87-9c7b-838ee58492ea.png)

hahmadia reviewed (changes requested):
Found some minor bugs to address.

Note I didn't test every single property since some have overlap but you should check them all just to make sure I didn't miss something by accident. Thanks

harshilsharma63 commented:
Thanks @hahmadia . I'm working on these.

harshilsharma63 commented:
> When the column too small, the range gets stacked on top like this ( I think it should be come ellipsis)

Fixed.

> For date, the unique should be 3 here right?

Hmm, can you share an export of the board? Also can you verify it on the latest branch? I think it's fixed when I merged `main` in.

> Count values for multi select should be 7 here right?

Fixed

> For Checkbox, shouldn't false values be accounted for too?

Later on, we'll add checkbox specific values - `Checked` and `Unchecked`, those would make more sense. I think for now this is fine.

> Unique values here should be 2 (I assume we account for false values too if I'm not wrong)

Same as considering false values, later on, we'll add checkbox specific values - `Checked` and `Unchecked`, those would make more sense. I think for now this is fine.

> Created Time 0 values? Is this right?

Fixed

> Created time unique values also 0?

Fixed

> Created by value is 0?

Fixed

> Created by unique value is also 0?

Fixed

hahmadia commented:
Hi Harshil,

Regarding todays meeting, I would be in favor of having this QA review after dev review.

harshilsharma63 commented:
@hahmadia you can review this now. I've synced this with `main`.

Agree on the QA point.

hahmadia commented:
Hi Harshil

Menu gets hidden on numbers. Also recommend adding "..." to range so we can know we should expand. I thought it was a bug but realized column was too small.

![image](https://user-images.githubusercontent.com/17804942/128354809-8c731324-8ea8-4d6a-bf45-116e51d7d3d0.png)

---

Have date like this

![image](https://user-images.githubusercontent.com/17804942/128355898-607574a6-60bc-4d27-96be-c4b206fd27ec.png)

Delete first value

![image](https://user-images.githubusercontent.com/17804942/128355957-184ef1bc-f56e-459d-83d9-c649f2776885.png)

Change first value to '08/13/2021'
![image](https://user-images.githubusercontent.com/17804942/128356023-9a8ef65d-84ac-4478-b24f-55b3d3bef19c.png)

Unique value stays at 2

**Changing it in the card using the calendar works fine and doesn't cause this problem**

---

Shouldn't unique only be 4 for created time here? 

![image](https://user-images.githubusercontent.com/17804942/128356807-3fe6746d-f692-4d99-89e4-1ac05ca5c452.png)

---

Last updated time unique should be 2 here right?
![image](https://user-images.githubusercontent.com/17804942/128357200-74d53f74-24d3-47a7-8c21-f2e0ad43e410.png)



Otherwise nice job! Everything else seems to be working. I won't wait for you to fix these to not hold you back. I'll go forward with doing code review now.

hahmadia commented on webapp/src/components/calculations/calculations.ts:
```diff
@@ -0,0 +1,180 @@
+import {Utils} from '../../utils'
+import {Constants} from '../../constants'
+
+const roundedDecimalPlaces = 2
```
Probably wanna make this all caps

![image](https://user-images.githubusercontent.com/17804942/128706491-18663989-cf76-4f4d-a499-8662fe523a3a.png)

hahmadia reviewed (approved):
Looks good to me with the exception of incorrect calculations

ogi-m commented:
@hahmadia  Regarding the unique count on the dates:
If there's 3 dates in the column - 2x`August 13` and 1x`August 31` and `Unique` count shows 2 - isn't that correct?

hahmadia commented:
> @hahmadia  Regarding the unique count on the dates:
> 
> If there's 3 dates in the column - 2x`August 13` and 1x`August 31` and `Unique` count shows 2 - isn't that correct?

@ogi-m 

Yes sorry! That is correct.

hahmadia commented:
/cherry-pick release-0.9.0

mattermod commented:
Cherry pick is scheduled.

mattermod commented:
Error trying doing the automated Cherry picking. Please do this manually

```

```

chenilim commented:
@metanerd (I think you still own mattermod?), do you know if slash cherry-pick should work on this repo? If so, is there something that needs to be configured? Thanks.
